**import the required libraries.**

In [1]:
# import 'pandas' 
import pandas as pd 

# import 'numpy' 
import numpy as np

# to suppress warnings 
from warnings import filterwarnings
filterwarnings('ignore')

# import 'random' to generate random sample
import random

# import statistics to perform statistical computation  
import statistics

# import 'stats' package from scipy library
from scipy import stats

# import a library to perform Z-test
from statsmodels.stats import weightstats as stests

# to test the normality 
from scipy.stats import shapiro

# import the function to calculate the power of test
from statsmodels.stats import power

### 1. Point Estimation

#### 1. Consider the data of grade points for 35 students in a data science course. Select grades of 20 students randomly from the data and find the point estimate for the population mean.

     Grades: [59.1, 65.0, 75.8, 79.2, 95.0, 99.8, 89.1, 65.2, 41.9, 55.2, 94.8, 84.1, 83.2, 74.0, 75.5, 76.2, 79.1, 80.1, 
              92.1, 74.2, 59.2, 64.0, 75, 78.2, 95.6, 97.8, 89.5, 64.2, 41.8, 57.2, 85, 91.4, 81.8, 74.6, 90]

In [2]:
Grades = [59.1, 65.0, 75.8, 79.2, 95.0, 99.8, 89.1, 65.2, 41.9, 55.2, 94.8, 84.1, 83.2, 74.0, 75.5, 76.2, 79.1, 80.1, 
          92.1, 74.2, 59.2, 64.0, 75, 78.2, 95.6, 97.8, 89.5, 64.2, 41.8, 57.2, 85, 91.4, 81.8, 74.6, 90]
    
# radom seed  = 1 ensures the same sample is drawn each time the code is run    
random.seed(1)

sample_grades = random.sample(population = Grades, k = 20)

sample_mean = np.mean(sample_grades)

print("The point estimate for the population mean is",round(sample_mean,2))

The point estimate for the population mean is 75.33


#### 2. A financial firm has created 50 portfolios. From them, a sample of 13 portfolios was selected, out of which 8 were found to be underperforming. Can you estimate the number of underperforming portfolios?

In [3]:
# Total count of portfolios
N = 50

# Number of sample portfolios
n = 13

# Number of underperforming portfolios out of the sample
x = 8

# Sample Proportion
p_samp = x/n

# Estimate Number of unerperforming portfolios
est_num_port = N*p_samp

print("The number of underperforming portfolios are",round(est_num_port))

The number of underperforming portfolios are 31


### 2. Sampling Error

#### 1. Consider the data for the number of ice-creams sold per day. An ice-cream vendor collected this data for 90 days and then a sample is drawn (without replacement) containing ice-creams sold for 25 days. 

data = [21, 93, 62, 76, 73, 20, 56, 95, 41, 36, 38, 13, 80, 88, 34, 18, 40, 11, 
        25, 29, 61, 23, 82, 10, 92, 69, 60, 87, 14, 91, 94, 49, 57, 83, 96, 55, 
        79, 52, 59, 39, 58, 17, 19, 98, 15, 54, 48, 46, 72, 45, 65, 28, 37, 30, 
        68, 75, 16, 33, 31, 99, 22, 51, 27, 67, 85, 47, 44, 77, 64, 97, 84, 42, 
        90, 70, 74, 89, 32, 26, 24, 12, 81, 53, 50, 35, 71, 63, 43, 86, 78, 66]
        
sample = [10, 22, 47, 66, 11, 57, 77, 98, 31, 63, 74, 84, 50, 96, 88, 92, 70, 54, 65, 44, 16, 72, 20, 90, 43]

Comupte the sampling error for mean.

In [4]:
# Population
data = [21, 93, 62, 76, 73, 20, 56, 95, 41, 36, 38, 13, 80, 88, 34, 18, 40, 11, 25, 29, 61, 23, 82, 10, 92, 69, 60, 87, 14, 91, 94, 49, 57, 83, 96, 55, 79, 52, 59, 39, 58, 17, 19, 98, 15, 54, 48, 46, 72, 45, 65, 28, 37, 30, 68, 75, 16, 33, 31, 99, 22, 51, 27, 67, 85, 47, 44, 77, 64, 97, 84, 42, 90, 70, 74, 89, 32, 26, 24, 12, 81, 53, 50, 35, 71, 63, 43, 86, 78, 66]

# Sample
sample = [10, 22, 47, 66, 11, 57, 77, 98, 31, 63, 74, 84, 50, 96, 88, 92, 70, 54, 65, 44, 16, 72, 20, 90, 43]

In [5]:
# Calculation of point estimate for the population mean
samp_mean = np.mean(sample)

# Calculation of population mean
pop_mean = np.mean(data)
pop_std = np.std(data)

# calculation of sampling error for mean
print("Sampling error for mean is",round(np.abs(pop_mean-samp_mean),2))

Sampling error for mean is 3.1


### 3. Interval Estimation for Mean

#### 1. A random sample of weight (in kg.) for 35 diabetic patients is drawn from the population with a standard deviation of 8 kg. Find the 90% confidence interval for the population mean.

    Weight: [59.1, 65.0, 75.8, 79.2, 95.0, 99.8, 89.1, 65.3, 41.9, 55.2, 94.8, 84.1, 83.2, 74.0, 75.5, 76.2, 79.1, 80.1, 
             92.1, 74.2, 59.2, 64.0, 75, 78.2, 95.6, 97.8, 89.5, 64.2, 41.8, 57.2, 85, 91.4, 81.8, 74.6, 90]

In [6]:
Weight = [59.1, 65.0, 75.8, 79.2, 95.0, 99.8, 89.1, 65.3, 41.9, 55.2, 94.8, 84.1, 83.2, 74.0, 75.5, 76.2, 79.1, 80.1, 
         92.1, 74.2, 59.2, 64.0, 75, 78.2, 95.6, 97.8, 89.5, 64.2, 41.8, 57.2, 85, 91.4, 81.8, 74.6, 90]

In [7]:
n = len(Weight)

samp_std = 8

# Confidence interval to be taken is 90%
ci = stats.norm.interval(0.90, loc = np.mean(Weight), scale = samp_std/np.sqrt(n))

print("The 90% confidence interval of population mean is",np.round(ci,2))

The 90% confidence interval of population mean is [74.46 78.91]


#### 2. There are 150 apples on a tree. You randomly choose 40 apples and found that the average weight of apples is 182 grams with a standard deviation of 30 grams. Find the 95% confidence interval for the population mean.

In [8]:
N = 150

# Sample
n = 40

# sample mean
samp_mean = 182

# sample standard deviation
samp_std = 30

ci_95 = stats.norm.interval(0.95, loc = 182, scale = samp_std/np.sqrt(n))

print("The 95% confidence interval of population mean is",np.round(ci_95,2))

The 95% confidence interval of population mean is [172.7 191.3]


#### 3. A movie production house needs to estimate the average monthly wage of the technical crew members. The previous data shows that the standard deviation of the wages is 190 dollars. The production team thinks that the estimation of the average wage should not exceed 54 dollars. The team has decided to take a small subset of wages for the estimation. Find a suitable number of wages to be considered to get the estimate with 90% confidence.

In [9]:
# population standard deviation
sigma = 190

# Required Margin of error
ME = 54

# calculate z_alpha_by_2 with alpha  = 0.1
z_alpha_by_2 = np.abs(round(stats.norm.isf(0.1/2),4))

# calculate sample size (n)
n = (z_alpha_by_2**2)*((sigma**2)/(ME**2))

print("The required sample size is",round(n))

The required sample size is 33


#### 4. 100 bags of coal were tested and had an average of 35% of ash with a standard deviation of 15%. Calculate the margin of error for a 90% confidence level.

In [10]:
# ME = (Z_alpha_by_2)*(sample_std/sqrt(n))

n = 100
sample_mean = 0.35
sample_std = 0.15

# Confidence interval to be taken is 90%

z_alpha_by_2 = round(stats.norm.isf(0.1/2),4)

ME = (z_alpha_by_2)*(sample_std/np.sqrt(n))

print("Margin of error is",round(ME,4),"or",round(ME*100,2))

Margin of error is 0.0247 or 2.47


#### 5. From a sample of 250 observations, it is found that the average income of a 27 year old Londoner is £45,000 with a sample standard deviation of £4000. Obtain the 95% confidence interval to estimate the average income.

In [11]:
n = 250
sample_mean = 45000
sample_std = 4000

# Confidence interval to be taken is 95%
ci_95_avg_inc = stats.norm.interval(0.95, loc = sample_mean, scale = sample_std/np.sqrt(n))

print("95% confidence interval for average income is",np.round(ci_95_avg_inc,4))

95% confidence interval for average income is [44504.164 45495.836]


#### 6. There are 150 apples on a tree. You randomly choose 17 apples and found that the average weight of apples is 78 grams with a standard deviation of 23 grams. Find the 90% confidence interval for the population mean.

In [12]:
N = 150

n = 17

sample_mean = 78
sample_std = 23

ci_90pct = stats.t.interval(0.90, df = n-1, loc = sample_mean, scale = sample_std/np.sqrt(n))

print("90% confidence interval for the population is",np.round(ci_90pct,2))

90% confidence interval for the population is [68.26 87.74]


#### 7. A financial firm has created 50 portfolios. From them, a sample of 13 portfolios was selected, out of which 8 were found to be underperforming. Construct a 99% confidence interval to estimate the population proportion.

In [13]:
# Total number of portfolios
N = 50

# Sample number of portfolios
n = 13

# number of underperforming portfolios
x = 8

# sample proportion
p_samp = x/n

ci = stats.norm.interval(0.99, loc = p_samp, scale = np.sqrt((p_samp*(1-p_samp))/n))

print("99% confidence interval for population proportion is",np.round(ci,2))

99% confidence interval for population proportion is [0.27 0.96]


### 4. Test of Hypothesis

#### 1. A car manufacturing company claims that the mileage of their new car is 25 kmph with a standard deviation of 2.5 kmph. A random sample of 45 cars was drawn and recorded their mileage as per the standard procedure. From the sample, the mean mileage was seen to be 24 kmph. Is this evidence to claim that the mean mileage is different from 25kmph? (assume the normality of the data) Use α = 0.01.

The null and alternative hypothesis is:

H<sub>0</sub>: $\mu = 25 $<br>
H<sub>1</sub>: $\mu ≠ 25 $

Here ⍺ = 0.01, for a two-tailed test calculate the critical z-value.

In [14]:
z_val = np.abs(round(stats.norm.isf(q = 0.01/2),2))
print('Critical value for two tailed z-test is',z_val)

Critical value for two tailed z-test is 2.58


i.e. if z is less than -2.58 or greater than 2.58 then we reject the null hypothesis.

In [15]:
# calculate z-score

def z_calc(pop_mean, pop_std, n, samp_mean):
    z_score = (samp_mean - pop_mean)/(pop_std / np.sqrt(n))
    return z_score

n = 45
pop_mean = 25
pop_std = 2.5
samp_mean = 24

z_score = z_calc(pop_mean, pop_std, n, samp_mean)
print('Z-Score is',z_score)

Z-Score is -2.6832815729997477


In [16]:
# calculation of p-value
# We use .sf - survival function to calculate the p-value at the tails using absolute value of z-score
p_val = stats.norm.sf(abs(z_score)) # two-tailed test

# Since it is a two-tailed test we multiply p-value by 2
req_p_value = p_val*2

print("p-value is",req_p_value)

p-value is 0.007290358091535638


Hence,  p-value is less than 0.01 we reject Null Hypothesis. Hence, we conclude that there is enough evidence to claim that the mean mileage is different from 25kmph.

#### 2. The average calories in a slice bread of the brand 'Alphas' are 82 with a standard deviation of 15. An experiment is conducted to test the claim of the dietitians that the calories in a slice of bread are not as per the manufacturer's specification. A sample of 40 slices of bread is taken and the mean calories recorded are 95. Test the claim of dietitians with ⍺ value (significance level) as 0.05. (assume the normality of the data).

The null and alternative hypothesis is:

H<sub>0</sub>: $\mu = 82 $<br>
H<sub>1</sub>: $\mu ≠ 82 $

Here ⍺ = 0.05, for a two-tailed test calculate the critical z-value.

In [17]:
pop_mean = 82
pop_std = 15

n = 40
samp_mean = 95

In [18]:
z_val = np.abs(round(stats.norm.isf(q = 0.05/2),2))
print("Critical value for two-tailed test is",z_val)

Critical value for two-tailed test is 1.96


if z is less than -1.96 or greater than 1.96 then we reject the null hypothesis.

In [19]:
z_score = z_calc(pop_mean, pop_std, n, samp_mean)
print("Z-Score",z_score)

Z-Score 5.4812812776251905


In [20]:
# calculation of p-value
# We use .sf - survival function to calculate the p-value at the tails using absolute value of z-score
p_val = stats.norm.sf(abs(z_score)) # two-tailed test

# Since it is a two-tailed test we multiply p-value by 2
req_p_value = p_val*2

print("p-value is",req_p_value)

p-value is 4.222565249683579e-08


As p-value is less than 0.05 we reject null hypothesis. Thereby, we conclude that there is enough evidence to average calories in a slice of bread are not 82.

#### 3. A typhoid vaccine in the market inscribes 3 mg of ascorbic acid in the vaccine. A research team claims that the vaccines contain less than 3 mg of acid. We collected the data of 40 vaccines by using random sampling from a population and recorded the amount of ascorbic acid. Test the claim of the research team using the sample data ⍺ value (significance level) to 0.05.

    acid_amt = [2.57, 3.06, 3.28 , 3.24, 2.79, 3.40, 3.36, 3.07, 2.46, 3.03, 3.05, 2.94, 3.46, 3.19, 3.09, 2.81, 3.13, 2.88, 
                2.76, 2.75, 3.17, 2.89, 2.54, 3.18, 3.08, 2.60, 3.06, 3.13, 3.11, 3.08, 2.93, 2.90, 3.06, 2.97, 3.24, 2.86, 
                2.87, 3.18, 3, 2.95]

In [21]:
acid_amt = [2.57, 3.06, 3.28 , 3.24, 2.79, 3.40, 3.36, 3.07, 2.46, 3.03, 3.05, 2.94, 3.46, 3.19, 3.09, 2.81, 3.13, 2.88, 
            2.76, 2.75, 3.17, 2.89, 2.54, 3.18, 3.08, 2.60, 3.06, 3.13, 3.11, 3.08, 2.93, 2.90, 3.06, 2.97, 3.24, 2.86, 
            2.87, 3.18, 3, 2.95]

Let us check the normality of the data.

Ho: Data is normal

Ha: Data is not normal

In [22]:
# Performing shapiro-Wilk test to test normality
stat, p_val = shapiro(acid_amt)

# print test statistic and p-value
print("Test statistic",stat)
print("P-value",p_val)

Test statistic 0.9764794111251831
P-value 0.5609433054924011


From the above result, we can see that the p-value is greater than 0.05, we fail to reject null hypothesis thus we can say that the data is normally distributed.

The null and alternative hypothesis is:

H<sub>0</sub>: $\mu \geq 3 $<br>
H<sub>1</sub>: $\mu < 3 $

Here ⍺ = 0.05, for a one-tailed test calculate the critical z-value.

In [23]:
z_val = np.abs(stats.norm.isf(q = 0.05))
print("Critical value for One-tailed test",round(z_val,2))

Critical value for one-tailed test 1.64


Since we test for less than value | if z is less than -1.64 then we reject the null hypothesis.

In [24]:
z_score, p_val = stests.ztest(x1 = acid_amt, value = 3, alternative = 'smaller')
print("Z-Score",z_score)
print("P-Value",p_val)

Z-Score 0.08289008952836197
P-Value 0.5330305328333862


In [25]:
ci = stats.norm.interval(0.95, loc = np.mean(acid_amt), scale = statistics.stdev(acid_amt)/np.sqrt(len(acid_amt)))
print("Confidence Interval",np.round(ci,2))

Confidence Interval [2.93 3.07]


Thus, Z-score is greater than -1.64 and p-value is greater than 0.05, we fail to reject null hypothesis. Thus, there is no enough evidence to conlcude that the vaccine contains less than 3mg of acid.

#### 4. A sample of 900 PVC pipes is found to have an average thickness of 12.5 mm. Can we assume that the sample is coming from a normal population with mean 13 mm against that it is less than 13 mm. The population standard deviation is 1 mm. Test the hypothesis at 5% level of significance.

The null and alternative hypothesis is:

H<sub>0</sub>: $\mu \geq 13 $<br>
H<sub>1</sub>: $\mu < 13 $

Here ⍺ = 0.05, for a one-tailed test calculate the critical z-value.

In [26]:
n = 900
samp_mean = 12.5
pop_mean = 13
pop_std = 1

In [27]:
z_val = np.abs(round(stats.norm.isf(q = 0.05),2))
print("Critical value for One-tailed test is",z_val)

Critical value for two-tailed test is 1.64


In [28]:
z_score = z_calc(pop_mean, pop_std, n, samp_mean)
print("Z-Score",z_score)

Z-Score -15.0


In [29]:
# calculation of p-value
p_val = stats.norm.cdf(z_score)

print("p-value is",p_val)

p-value is 3.6709661993126986e-51


In [30]:
ci = stats.norm.interval(0.95, loc = samp_mean, scale = pop_std / np.sqrt(n))
print('Confidence interval:', np.round(ci,2))

Confidence interval: [12.43 12.57]


Here the z-score is less than -1.64, p-value is less than 0.05, also the confidence interval do not contain the value in the null hypothesis, thus we reject the null hypothesis and that there is no enough evidence to conclude that sample is coming from the population with mean 13.

#### 5. An e-commerce company claims that the mean delivery time of food items on its website in NYC is 60 minutes with a standard deviation of 30 minutes. A random sample of 45 customers ordered from the website, and the average time for delivery was found to be 75 minutes. Is this enough evidence to claim that the average time to get items delivered is more than 60 minutes. (assume the normality of the data). Test the with α = 0.05.

The null and alternative hypothesis is:

H<sub>0</sub>: $\mu \leq  60 $<br>
H<sub>1</sub>: $\mu > 60 $

Here ⍺ = 0.05, for a one-tailed test calculate the critical z-value.

In [31]:
z_val = np.abs(round(stats.norm.isf(q = 0.05), 2))

print('Critical value for one-tailed Z-test:', z_val)

Critical value for one-tailed Z-test: 1.64


if z is greater than 1.64 then we reject the null hypothesis.

In [32]:
n = 45
pop_mean = 60
pop_std = 30
samp_mean = 75

In [33]:
z_score = z_calc(pop_mean, pop_std, n, samp_mean)
print("Z-score:", z_score)

Z-score: 3.3541019662496843


In [34]:
ci =  stats.norm.interval(0.95, loc = samp_mean, scale = pop_std / np.sqrt(n))
print('Confidence interval is',np.round(ci,2))

Confidence interval is [66.23 83.77]


In this scenario, the z-score is greater than 1.64, the p-value is less than 0.05, also the confidence interval do not contain the value in the null hypothesis, thus we reject the null hypothesis and have enough evidence to conclude that the average time to get items delivered is more than 60 minutes.

### 5. Power of Hypothesis Test

The probability of correctly rejecting a false null hypothesis is defined as the `Power` of a test. It is calculated as `1-β`. It is the probability of avoiding the type II error. The power of the test lies between 0 to 1. The value close to 1 indicates that the test is able to reject the false null hypothesis correctly.

#### 1. We need to determine if girls' score on average is higher than 300 in the GRE (for higher studies). The standard deviation of the population is 20. We collected the data of 40 girls by using random sampling and recorded their marks. Set ⍺ value (significance level) to 0.05 to calculate the power of the test.

    GRE scores = [311, 313, 309, 307, 302, 313, 305, 336, 322, 300, 335, 312, 282, 314, 310, 316, 301, 315, 299, 298, 308, 
                  315, 306, 321, 276, 306, 324, 308, 334, 277, 319, 322, 298, 300, 293, 333,305, 302, 317, 287]

Let us check the normality of the data.

In [35]:
GRE_scores = [311, 313, 309, 307, 302, 313, 305, 336, 322, 300, 335, 312, 282, 314, 310, 316, 301, 315, 299, 298, 308, 
              315, 306, 321, 276, 306, 324, 308, 334, 277, 319, 322, 298, 300, 293, 333,305, 302, 317, 287]

In [36]:
stat, p_value = shapiro(GRE_scores)
print('Test statistic:', stat)
print('P-Value:', p_value)

Test statistic: 0.9695490598678589
P-Value: 0.3482360541820526


From the above result, we can see that the p-value is greater than 0.05, thus we can say that the data is normally distributed.

The null and alternative hypothesis is:

H<sub>0</sub>: $\mu \leq 300$<br>
H<sub>1</sub>: $\mu > 300$

In [37]:
z_val = np.abs(round(stats.norm.isf(q = 0.05),2))

print('Critical value for one-tailed Z-test:', z_val)

Critical value for one-tailed Z-test: 1.64


In [38]:
# number of observations 
num_obs = len(GRE_scores)

# calculate the sample mean
samp_mean = np.mean(GRE_scores)

# null hypothesis value
pop_mean = 300

# population standard deviation
pop_std = 20

b = pop_mean + ((z_val)*(pop_std/np.sqrt(num_obs)))
print(b)

305.18613536267617


In [39]:
pwr = power.zt_ind_solve_power(effect_size = (samp_mean-pop_mean)/pop_std,
                        nobs1 = num_obs,
                        alpha = 0.05,
                        power = None,
                        ratio = 0,
                        alternative = 'larger')

print("Power of test is",pwr)

Power of test is 0.8707713725498709


The power of test is 0.87. Thus, there is 87% chance that the test rejects a false null hypothesis. 
Higher value of power indicates the test is more powerful.